DigitalLearningSolutions.Data/Services/UserVerificationService.cs

```diff
@@ -0,0 +1,89 @@
+﻿namespace DigitalLearningSolutions.Data.Services
+{
+    using System.Collections.Generic;
+    using System.Linq;
+    using DigitalLearningSolutions.Data.DataServices.UserDataService;
+    using DigitalLearningSolutions.Data.Models.User;
+
+    public interface IUserVerificationService
+    {
+        UserAccountSet VerifyUsers(
+            string password,
+            List<AdminUser> unverifiedAdminUsers,
+            List<DelegateUser> unverifiedDelegateUsers
+        );
+
+        List<DelegateUser> GetVerifiedDelegateUsersAssociatedWithAdminUser(AdminUser? adminUser, string password);
+
+        AdminUser? GetVerifiedAdminUserAssociatedWithDelegateUsers(List<DelegateUser> delegateUsers, string password);
+    }
+
+    public class UserVerificationService : IUserVerificationService
+    {
+        private readonly ICryptoService cryptoService;
+        private readonly IUserDataService userDataService;
+
+        public UserVerificationService(ICryptoService cryptoService, IUserDataService userDataService)
+        {
+            this.cryptoService = cryptoService;
+            this.userDataService = userDataService;
+        }
+
+        public UserAccountSet VerifyUsers(
+            string password,
+            List<AdminUser> unverifiedAdminUsers,
+            List<DelegateUser> unverifiedDelegateUsers
+        )
+        {
+            var verifiedAdminUsers =
+                unverifiedAdminUsers.Where(au => cryptoService.VerifyHashedPassword(au.Password, password))
+                    .ToList();
+            var verifiedDelegateUsers =
+                unverifiedDelegateUsers.Where(du => cryptoService.VerifyHashedPassword(du.Password, password))
+                    .ToList();
+
+            return new UserAccountSet(verifiedAdminUsers, verifiedDelegateUsers);
+        }
+
+        public List<DelegateUser> GetVerifiedDelegateUsersAssociatedWithAdminUser(
+            AdminUser? adminUser,
+            string password
+        )
+        {
+            if (string.IsNullOrEmpty(adminUser?.EmailAddress))
+            {
+                return new List<DelegateUser>();
+            }
+
+            var delegatesAssociatedWithAdmin = userDataService.GetDelegateUsersByEmailAddress(adminUser.EmailAddress!);
+
+            var suitableDelegates = delegatesAssociatedWithAdmin
+                .Where(du => du.Active && du.Approved && cryptoService.VerifyHashedPassword(du.Password, password));
+
+            return suitableDelegates.ToList();
+        }
+
+        public AdminUser? GetVerifiedAdminUserAssociatedWithDelegateUsers(
+            List<DelegateUser> delegateUsers,
+            string password
+        )
+        {
+            var delegateEmail = delegateUsers.FirstOrDefault(du => du.EmailAddress != null)?.EmailAddress;

```

Comment:
Is taking the first email address fine? Is there a situation that can occur where we have multiple delegate users with different emails here? I don't think there is - the only time where getting associated admin users would differ from the initial user retrieval would be if a candidate number was used to log in, in which case there would only be one delegate user anyway. Just worth double checking.

Reply:
The only place this is called is from within the login service when getting linked accounts. If the initial verified account set has multiple emails, we're throwing an exception (including when there is 1 email + null). It might become a problem if other methods start calling the same method (I had to move it into this new service to get rid of a circular dependency) so I'll just add a comment above.